### ISSUE TITLE
Ricochet can lose messages

### ISSUE THREAD

**Issue by taoeffect** [ricochet-im/ricochet#282]
Just tested this with 1.1.1:

- Bob sends a message to Alice who is offline
- Bob goes offline
- Alice comes online
- Bob comes online
- Either Bob or Alice send additional messages

Original message is lost.

**Comment by taoeffect** [comment]
One simple way to do this (which would probably result in the closing of some other issues), is for Ricochet to keep encrypted logs of conversations (maybe add it as a checkbox option). It can then store, encrypted, an unsent message and continue re-trying next time the app is open.

More complicated ways would probably involve third-parties holding onto the messages. There can be a network of such escrow servers that can't read the messages, don't know who they belong to or where they're going, but through the magic of onion routing deliver them on time when the user goes back online.

Obviously such escrow servers would need to be paid for their services. Cryptocurrency could be integrated here. It should be integrated into the app anyway, but that's a long way down the road... (Working on it!)

**Comment by yawnbox** [comment]
I've experienced this too and see this as a positive feature. Any option to address this would fundamentally increase the potential of recovering communication content or metadata, which is contrary to the goals of this project.

When Alice and Bob are physically together and speaking directly to each other (and nothing is recording them), and Alive leaves Bob, Bob would be speaking to no one but himself if he kept talking. This is a privacy benefit, and anything to "fix" this would result in security consequences.

**Comment by taoeffect** [comment]
To not fix this problem would be to limit the usefulness of this application as a tool for secure communication. Limit its usefulness and few people will use it.

If the goal is for a small number of people to use this app, by all means, close this issue.

Ultimately some other app will come along that does address this problem well, and it will get recommended instead. People will use that app instead of this, and users will find few people who are willing to communicate with them using Ricochet.

I was under the impression that the goal was to create a usable, user-friendly secure and private communications tool.

**Comment by taoeffect** [comment]
I mean, if this is a big concern for Ricochet's user base (not clear that it is), then the "simple method" I suggested above can be implemented in such a way that does not compromise anyone's privacy.

Let note another issue first:

If someone gets ahold of my computer and are able to open Ricochet they can pretend to be me. There is no security here for that problem. No password. That's not good as it means that the person you're talking with could be setting a trap for you, making you think that you're talking with someone else in reality.

Adding a password to encrypt the private key that's used to sign messages would solve that problem.

Bringing us back to this issue: adding a password would also allow for the option to encrypt unsent messages upon quit. You wouldn't even have to store logs of conversations if you didn't want to (though that would be a very convenient optional feature), just the unsent messages until the next time you open and unlock the program.

The magical relay network can come later, maybe even in a different app.

**Comment by s-rah** [comment]
This is something that has bitten me once or twice recently. And I am torn on the solution.

On the one hand I see the argument here. Ricochet should cache unsent messages locally. This one seems the nice from a  privacy and UX perspective - and there is precedent for local storage. Personally I don't see a huge issue including this - as long as it is signposted and the messages are encrypted locally (something that we should really be doing for the ricochet config file anyway).

Not a fan of 3rd parties, it adds so much new stuff to the threat model.

I understand @yawnbox concerns - which is why I think we should limit this to unsent local messages. There have been enough usecases where logs have bitten people that it is definitely worth considering whether this should be default .

But then,  most people use IM's for real time, context sensitive communication. Often if someone goes offline and reappears the next day the context is gone and there is no need for the message to continue. This makes me wonder about the overall need for such a feature  - how often do unsent messages need to be sent after the fact? I would put money on it being low. If it is very important, the message will be sent again, or over a more permanent medium - like email or pond.

I think I probably fall down on the side of "drop the message" (or maybe, in some future, offer an option to send a pond message to the user if they are offline) - for most usecases, caching the messages is unlikely to be wanted behavior.

**Comment by taoeffect** [comment]
There's another option that I think should be wholly uncontroversial: make it clear that quitting means the message will be lost.

There's already a huge problem with the UI being too subtle to give useful user feedback. "White" means unsent, "grey" means received. That's pretty subtle. Most users won't pick up on that difference and its meanings.

So at the *very least:*:

1. Add a "Received" or "Delivered" message in tiny font below the most recently delivered message (similar to how Signal works).
2. Add a "Are you sure you want to quit? Unsent messages will be lost" dialog when the user quits with undelivered messages.

**Comment by special** [comment]
I'm not opposed to having the ability to save unsent messages locally, as long as it's done the right way. In particular:

1. There's a dependency on encrypted local storage (#33).
2. Messages should only be saved _at exit_ if unsent, and probably with a confirmation dialog (at least the first time). That means they'll still be lost if Ricochet exits unexpectedly, but many fewer messages hit the disk.
3. There should be a way to delete an unsent message through UI

The confirmation dialog on exit is a good, uncomplicated step.

I bet many of the people who want this will also want to save some recent messages for context after restarting Ricochet. Hmm.

**Comment by taoeffect** [comment]
@special What do you think about my comments above about also using the password to protect people's identity by using it to encrypt the private key that's used for signing messages?

**Comment by special** [comment]
@taoeffect see #33 - this wasn't possible until tor implemented ADD_ONION. Now it's just waiting on an implementation.

**Comment by PacoBell** [comment]
Does this project have any ambitions of operating on mobile? Because that environment is typically async, unstable connectivity, and the app can be killed at any time. It's one of the reasons why Signal went with third-party servers.

**Comment by Mromson** [comment]
I think giving the option of keeping a chatlog stored locally is extremely important for any IM. A good idea would probably to give the user a prompt on first startup whether he/she wishes for the client to store chatlog and/or unsent messages. Right now, if the client is compromised, the user is screwed anyways - and being able to keep a chatlog / send messages to offline users is a top priority for most people.

**Comment by special** [comment]
@Mromson There's a previous discussion on logging in #168. That would be a better place to discuss chat logs, as opposed to persisting unsent messages.

>  Right now, if the client is compromised, the user is screwed anyways

There are different types of "screwed". Having the contact list or being able to impersonate a person is different from being able to forensically gather past conversation history.
